src/proc.cpp

```diff
@@ -395,35 +428,60 @@ static bool process_mark_finished_children(bool block_on_fg) {
         // wait on only one pgrp at a time and we need to check all pgrps before returning, but we
         // never wait/block on fg processes after an error has been encountered to give ourselves
         // (elsewhere) a chance to handle the fallout from process termination, etc.
-        if (!has_error && block_on_fg && j->pgid != shell_pgid
-                && j == job_fg && j->get_flag(job_flag_t::JOB_CONTROL)) {
+        if (!has_error && block_on_fg && j->pgid != shell_pgid && j == job_fg
+                && j->get_flag(job_flag_t::JOB_CONTROL)) {
             debug(4, "Waiting on processes from foreground job %d", job_fg->pgid);
             options &= ~WNOHANG;
         }
 
-        int status = -1;
-        // A negative PID passed in to `waitpid()` means wait on any child in that process group
-        auto pid = waitpid(-1 * j->pgid, &status, options);
-        if (pid > 0) {
-            // A child process has been reaped
-            handle_child_status(pid, status);
-            return true;
-        }
-        else if (pid == 0) {
-            // No killed/dead children in this particular process group
-        } else {
-            // pid < 0 indicates an error. One likely failure is ECHILD (no children), which is not
-            // an error and is ignored. The other likely failure is EINTR, which means we got a
-            // signal, which is considered an error. We absolutely do not break or return on error,
-            // as we need to iterate over all constructed jobs but we only call waitpid for one pgrp
-            // at a time. We do bypass future waits in case of error, however.
-            if (errno != ECHILD) {
-                has_error = true;
-                wperror(L"waitpid in process_mark_finished_children");
+        // waitpid(2) returns 1 process each time, we need to keep calling it until we've reaped all
+        // children of the pgrp in question or else we can't reset the dirty_state flag. In all
+        // cases, calling waitpid(2) is faster than potentially calling select_try() on a process
+        // that has exited, which will force us to wait the full timeout before coming back here and
+        // calling waitpid() again.
+        while (true) {
```

Comment:
Isn't this a bit clearer as `while (auto pid = waitpid`?

Reply:
(I'm not a massive fan of `while (true)` because that means you need to manually account for all possibilities to not have an infinite loop)

Reply:
Changed. 

rust's preference for the unconditional `loop` over `while ...` has gotten to me, and I found it more logical to explicitly "handle" all possible results of the `waitpid` call instead of split the handling across the while condition and inside the body.

Reply:
AFAIK rust has compiler support for checking that you handled all possible results. This is C++, which... doesn't.

Reply:
Very fair point!